### PR TITLE
Add test for factories

### PR DESCRIPTION
**Pull request by BernhardPosselt**
In response to #17755 and @Xenopathic 's comment 

Turns out our non shared factories don't work xD

Any ideas @DeepDiver1975 @MorrisJobke @PVince81 @rullzer

### CODE REVIEW

File: tests/lib/appframework/utility/SimpleContainerTest.php [hunk 1]
```diff
@@ -176,5 +176,25 @@ public function testConstructorComplexNoTestParameterFound() {
...
+
+    public function testRegisterAliasFactory() {
+        $counter = 0;
+        $this->container->registerService('test', function() use ($counter) {
+            $counter += 1;
+            return $counter;
+        }, false);
+        $this->container->registerAlias('test1', 'test');
+        $this->assertEquals(1, $this->container->query('test1'));
+        $this->assertEquals(2, $this->container->query('test1'));
```
Comment (Xenopathic) [inline tests/lib/appframework/utility/SimpleContainerTest.php]: Query the original service, `test`, as well to make sure the 'instances' are the same.

Comment (Raydiation) [inline tests/lib/appframework/utility/SimpleContainerTest.php]: Probably does not make sense given it's a scalar value

Comment (Raydiation) [inline tests/lib/appframework/utility/SimpleContainerTest.php]: That is updated every time you access it ;)

**Comment by Xenopathic** [comment]
Turns out in DIContainer and Server we didn't call `parent::__construct()`, so the Pimple\Container never initialized its internal members. Be a good OOP citizen guys, remember to construct your parents!

I tweaked the tests to use an object, which we can check for equivalence with the PHPUnit `$this->assertSame()` check. Less error-prone than counters :smile:

**Comment by Raydiation** [comment]
Awesome :) oh man I think this was a very important bugfix, we should probably back port it @karlitscheck

**Comment by Raydiation** [comment]
@karlitschek @DeepDiver1975

**Comment by Xenopathic** [comment]
@owncloud-bot retest this please

**Comment by Xenopathic** [comment]
Hmm, turns out that makes DIContainer have more state than the routing tests were expecting, so they fell over. Using a single DIContainer works fine though.

**Comment by Xenopathic** [comment]
@owncloud-bot retest this please

**Comment by Raydiation** [comment]
@owncloud-bot retest this please

**Comment by Raydiation** [comment]
reviews please 

cc @LukasReschke

**Comment by Raydiation** [comment]
@DeepDiver1975 any idea why it fails?

**Comment by Xenopathic** [comment]
@Raydiation The Jenkins slaves sometimes get confused when you rebase commits. Sometimes a retest fixes it, other times you need to create a fake PR with the commits (which can then be deleted later, as the status gets carried over to here).

**Comment by Xenopathic** [comment]
@owncloud-bot retest this please

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
@nickvergessen @icewind1991 please review - thx

**Comment by icewind1991** [comment]
Looks good :+1:

**Comment by Xenopathic** [comment]
@karlitschek Backport?

**Comment by karlitschek** [comment]
yes. please backport

**Comment by MorrisJobke** [comment]
@Raydiation @Xenopathic Can you take care of the backport?

**Comment by Xenopathic** [comment]
I'm on it

**Pull request merged**